File: src/main/java/kr/co/lunatalk/domain/exhibition/repository/ExhibitionRepositoryImpl.java
package kr.co.lunatalk.domain.exhibition.repository;

import com.querydsl.core.types.dsl.BooleanExpression;
import com.querydsl.jpa.impl.JPAQueryFactory;
import kr.co.lunatalk.domain.exhibition.domain.Exhibition;
import kr.co.lunatalk.domain.exhibition.domain.ExhibitionVisibility;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

import java.time.LocalDateTime;
import java.util.List;

import static kr.co.lunatalk.domain.exhibition.domain.QExhibition.exhibition;

@Repository
@RequiredArgsConstructor
public class ExhibitionRepositoryImpl implements ExhibitionRepositoryCustom {
	private final JPAQueryFactory queryFactory;

	@Override
	public List<Exhibition> findActiveExhibitions(ExhibitionVisibility visibility, LocalDateTime now) {
		return queryFactory
			.selectFrom(exhibition)
			.where(
				exhibition.visibility.eq(visibility)
					.and(exhibition.startAt.loe(now))
					.and(isActiveEndDate(now))
			)
			.fetch();
	}

	private BooleanExpression isActiveEndDate(LocalDateTime now) {
		return exhibition.endAt.isNull().or(exhibition.endAt.goe(now));
	}
}

File: src/main/kotlin/kr/co/lunatalk/domain/exhibition/repository/ExhibitionRepositoryImpl.kt
package kr.co.lunatalk.domain.exhibition.repository

import com.querydsl.core.types.dsl.BooleanExpression
import com.querydsl.jpa.impl.JPAQueryFactory
import kr.co.lunatalk.domain.exhibition.domain.Exhibition
import kr.co.lunatalk.domain.exhibition.domain.ExhibitionVisibility
import kr.co.lunatalk.domain.exhibition.domain.QExhibition.exhibition
import org.springframework.stereotype.Repository
import java.time.LocalDateTime

@Repository
class ExhibitionRepositoryImpl(
    private val queryFactory: JPAQueryFactory,
) : ExhibitionRepositoryCustom {

    override fun findActiveExhibitions(visibility: ExhibitionVisibility, now: LocalDateTime): List<Exhibition> {
        return queryFactory
            .selectFrom(exhibition)
            .where(
                exhibition.visibility.eq(visibility)
                    .and(exhibition.startAt.loe(now))
                    .and(isActiveEndDate(now))
            )
            .fetch()
    }

    private fun isActiveEndDate(now: LocalDateTime): BooleanExpression {
        return exhibition.endAt.isNull.or(exhibition.endAt.goe(now))
    }
}
